ENIGMAsystem/SHELL/Universal_System/Extensions/XInput/XIgamepad.cpp

```diff
@@ -17,238 +16,321 @@
+
+void extension_xinput_init() {
+  // Zero-initialize the gamepads array
+  for (size_t i = 0; i < XUSER_MAX_COUNT; ++i) {
+    auto& gamepad = gamepads[i];
+    gamepad.state = {};
+    gamepad.caps = {};
+    gamepad.gamepad_battery = {};
+  }
+  process_gamepads(); // update once for create/game start events
+  extension_update_hooks.push_back(process_gamepads);
+}
+
+} // namespace enigma
+
 namespace enigma_user {
 
 bool gamepad_is_supported() {
-	return true;
+  return true;
 }
-	
+
 bool gamepad_is_connected(int num) {
-	XINPUT_STATE controllerState;
-    // Zeroise the state
-    ZeroMemory(&controllerState, sizeof(XINPUT_STATE));
-    // Get the state
-    DWORD Result = XInputGetState(num, &controllerState);
-    if (Result == ERROR_SUCCESS) {
-        return true;
-    } else {
-        return false;
-    }
+  return (num >= 0 && num < XUSER_MAX_COUNT && gamepads[num].state_result == ERROR_SUCCESS);
 }
-	
+
 int gamepad_get_device_count() {
-	int count;
-	for (int i = 0; i < gamepad_get_max_device_count(); i++) {
-		count += gamepad_is_connected(i);
-	}
-	return count;
+  return XUSER_MAX_COUNT;
 }
 
-int gamepad_get_max_device_count() {
-	return XUSER_MAX_COUNT;
-}
-	
-std::string gamepad_get_description(int device) {
-	if (gamepad_is_connected(device)) {
-		return "Xbox 360 Controller (XInput STANDARD GAMEPAD)";
-	} else {
-		return "No device connected";
-	}
+string gamepad_get_description(int device) {
+  if (!gamepad_is_connected(device)) return "";
+  const auto& caps = gamepads[device].caps;
+
+  //if (caps.Type == XINPUT_DEVTYPE_GAMEPAD) {
+    switch (caps.SubType) {
+#ifdef XINPUT_DEVSUBTYPE_UNKNOWN
+      case XINPUT_DEVSUBTYPE_UNKNOWN:
+        break; // unknown
+#endif
+      case XINPUT_DEVSUBTYPE_GAMEPAD:
+        return "Xbox 360 Controller (XInput STANDARD GAMEPAD)";
+      case XINPUT_DEVSUBTYPE_WHEEL:
+        return "Xbox 360 Controller (XInput WHEEL)";
+      case XINPUT_DEVSUBTYPE_ARCADE_STICK:
+        return "Xbox 360 Controller (XInput ARCADE STICK)";
+#ifdef XINPUT_DEVSUBTYPE_FLIGHT_STICK
+      case XINPUT_DEVSUBTYPE_FLIGHT_STICK:
+        return "Xbox 360 Controller (XInput FLIGHT STICK)";
+#endif
+      case XINPUT_DEVSUBTYPE_DANCE_PAD:
+        return "Xbox 360 Controller (XInput DANCE PAD)";
+      case XINPUT_DEVSUBTYPE_GUITAR:
+        return "Xbox 360 Controller (XInput GUITAR)";
+#ifdef XINPUT_DEVSUBTYPE_GUITAR_ALTERNATE
+      case XINPUT_DEVSUBTYPE_GUITAR_ALTERNATE:
+        return "Xbox 360 Controller (XInput GUITAR ALTERNATE)";
+#endif
+#ifdef XINPUT_DEVSUBTYPE_GUITAR_BASS
+      case XINPUT_DEVSUBTYPE_GUITAR_BASS:
+        return "Xbox 360 Controller (XInput GUITAR BASS)";
+#endif
+      case XINPUT_DEVSUBTYPE_DRUM_KIT:
+        return "Xbox 360 Controller (XInput DRUM KIT)";
+#ifdef XINPUT_DEVSUBTYPE_ARCADE_PAD
+      case XINPUT_DEVSUBTYPE_ARCADE_PAD:
+        return "Xbox 360 Controller (XInput ARCADE PAD)";
+#endif
+    }
+    return "Xbox 360 Controller (XInput UNKNOWN)";
```

Comment:
What about non XBox controllers? Does XInput just not supply a mechanism for these?

Reply:
I don't think so, I think you have to install a special driver for a Sixaxis/DualShock to be recognized by XInput. That's probably why YoYo also loads 8 DirectInput gamepads.
https://docs2.yoyogames.com/source/_build/3_scripting/4_gml_reference/controls/gamepad%20input/index.html

>Windows is fully supported with up to a maximum of 12 connected devices permitted at once (numbered from 0 to 11, with 0 - 3 being XInput devices and 4 - 11 being DirectInput).

That said, I've heard Microsoft has been very good at getting XInput to recognize a lot of third party controllers, like Logitech. Anyway, we can come back to this later, but here's some more info about DirectInput interoperability with XInput:
https://docs.microsoft.com/en-us/windows/desktop/xinput/xinput-and-directinput